[SPARK-19239][PySpark] Check parameters whether equals None when specify the column in jdbc API

libratiger opened the pull request:
## What changes were proposed in this pull request?

The `jdbc` API do not check the `lowerBound` and `upperBound` when we
specified the ``column``, and just throw the following exception:

>```int() argument must be a string or a number, not 'NoneType'```


If we check the parameter, we can give a more friendly suggestion.


## How was this patch tested?
Test using the pyspark shell, without the lowerBound and upperBound parameters.

Commit:
[SPARK-19239][PySparK] Check the lowerBound and upperBound whether equal None in jdbc API

The ``jdbc`` API do not check the lowerBound and upperBound when we
specified the ``column``, and just throw the following exception:
```int() argument must be a string or a number, not 'NoneType'```
If we check the parameter, we can give a more friendly suggestion.

djvulee commented:
@zsxwing can you take a look at?

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented on python/pyspark/sql/readwriter.py:
```diff
@@ -431,6 +432,8 @@ def jdbc(self, url, table, column=None, lowerBound=None, upperBound=None, numPar
             if numPartitions is None:
                 numPartitions = self._spark._sc.defaultParallelism
+            assert lowerBound != None, "lowerBound can not be None when ``column`` is specified"
+            assert upperBound != None, "upperBound can not be None when ``column`` is specified"
```
Should we resemble the condition here - https://github.com/apache/spark/blob/55d528f2ba0ba689dbb881616d9436dc7958e943/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala#L100-L103 ?

djvulee replied:
Yes, The Scala code could check this, but the PySpark code will fail at ```int(lowerBound)``` first, so the customer is confused.

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #71472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71472/testReport)** for PR 16599 at commit [`94c44ba`](https://github.com/apache/spark/commit/94c44ba368acb3c7fa648ad66cfd3cac352af911).

SparkQA commented:
**[Test build #71472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71472/testReport)** for PR 16599 at commit [`94c44ba`](https://github.com/apache/spark/commit/94c44ba368acb3c7fa648ad66cfd3cac352af911).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71472/
Test FAILed.

Commit:
update the None equal comparison

SparkQA commented:
**[Test build #71476 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71476/testReport)** for PR 16599 at commit [`43602b5`](https://github.com/apache/spark/commit/43602b56d6099213a103a0c0389ac37ebb2c326b).

SparkQA commented:
**[Test build #71476 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71476/testReport)** for PR 16599 at commit [`43602b5`](https://github.com/apache/spark/commit/43602b56d6099213a103a0c0389ac37ebb2c326b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71476/
Test PASSed.

gatorsmile commented on python/pyspark/sql/readwriter.py:
```diff
@@ -431,6 +432,8 @@ def jdbc(self, url, table, column=None, lowerBound=None, upperBound=None, numPar
         if column is not None:
             if numPartitions is None:
                 numPartitions = self._spark._sc.defaultParallelism
```
This is contradicting with the scala version. Could you also change it to the following code
```Python
assert numPartitions is not None, "numPartitions can not be None when ``column`` is specified"
```

djvulee replied:
I have a little worry whether this change will break the API. If some users just specify the `column`, `lowerBound`, `upperBound` in some Spark version, its program will fail after update, even very few people just use the default  parallelism. 

In my personal opinion, I prefer to make a change and  keep API consistent.

If your opinion is to add the assert on `numPartitions`, I will update the PR soon.

gatorsmile replied:
I think we should make the Scala API and Python API consistent. The existing Python API is not following [the document](http://spark.apache.org/docs/2.1.0/sql-programming-guide.html). 
```
These options must all be specified if any of them is specified. They describe how to partition the table when reading in parallel from multiple workers. partitionColumn must be a numeric column from the table in question. Notice that lowerBound and upperBound are just used to decide the partition stride, not for filtering the rows in table. So all rows in the table will be partitioned and returned. This option applies only to reading.
```

gatorsmile commented:
Have you manually tested your code changes?

Commit:
Update numPartitions check to keep API consistent

In the Scala API, the `numPartitions` is needed when we specify the
`column`, we remove the default parallelism to keep consistent

SparkQA commented:
**[Test build #71488 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71488/testReport)** for PR 16599 at commit [`f2dad9c`](https://github.com/apache/spark/commit/f2dad9cdf77cab31152aab0f16df86cf004e41fe).

Commit:
Update the document for numPartition

djvulee commented:
I update the PR and test the change in pyspark shell.

SparkQA commented:
**[Test build #71490 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71490/testReport)** for PR 16599 at commit [`e9a1bca`](https://github.com/apache/spark/commit/e9a1bcad452dc231d185d1f4150f2a7a907f7ab7).

gatorsmile commented on python/pyspark/sql/readwriter.py:
```diff
@@ -429,8 +430,10 @@ def jdbc(self, url, table, column=None, lowerBound=None, upperBound=None, numPar
+            assert lowerBound is not None, "lowerBound can not be None when ``column`` is specified"
+            assert upperBound is not None, "upperBound can not be None when ``column`` is specified"
+            assert numPartitions is not None, "numPartitions can not be None " \
+                                              "when ``column`` is specified"
```
```Python
        assert numPartitions is not None, \
            "numPartitions can not be None when ``column`` is specified"
```

Commit:
Fix the python style

SparkQA commented:
**[Test build #71496 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71496/testReport)** for PR 16599 at commit [`d851382`](https://github.com/apache/spark/commit/d8513820d990d4235069a88249056b76bd35f60b).

SparkQA commented:
**[Test build #71488 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71488/testReport)** for PR 16599 at commit [`f2dad9c`](https://github.com/apache/spark/commit/f2dad9cdf77cab31152aab0f16df86cf004e41fe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71488/
Test PASSed.

SparkQA commented:
**[Test build #71490 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71490/testReport)** for PR 16599 at commit [`e9a1bca`](https://github.com/apache/spark/commit/e9a1bcad452dc231d185d1f4150f2a7a907f7ab7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71490/
Test PASSed.

SparkQA commented:
**[Test build #71496 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71496/testReport)** for PR 16599 at commit [`d851382`](https://github.com/apache/spark/commit/d8513820d990d4235069a88249056b76bd35f60b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71496/
Test PASSed.

gatorsmile commented:
LGTM

gatorsmile commented:
Thanks! Merging to master
